### DISCUSSION TITLE
MP3 decoder / CircuitPython Envy

### DISCUSSION THREAD

**Discussion by Guustaaf** [micropython/micropython#11307] [General]
CircuitPython has this great looking MP3 module.

MicroPython needs this too!

Calling all C++ geniuses!

My Picos with DACs are awaiting your heroic efforts. 🥇 

Thanks!

Guustaaf

**Comment by peterhinch** [comment]
A search on MP3 in the Adafruit site yields solutions based on the VS1053b chip. There is already a MicroPython driver for this [here](https://github.com/peterhinch/micropython-vs1053).

A software-only driver powering an I2S output would be nice, but I'm not aware of such a thing.

**Comment by Guustaaf** [comment]
Thanks for that Peter. I ordered a few breakout boards with that chip and will check it out. And thanks for making the driver for it!

**Comment by rkompass** [comment]
If I understand it right then Circuitpython has a module that enables mp3 decoding _natively_, the source is [here](https://github.com/adafruit/Adafruit_MP3/tree/7a5de1ad777e95b0f4fab7bbd35678c7d319b1b5).
This could/should be ported to Micropython by someone knowledgeable enough (I'm afraid it exceeds my expertise).
With this one would only need an amplifier to play sound from a file, given we have DAC/PWM and perhaps DMA working.
So I think the title is justified here. I would be eager to contribute to the "surrounding" Micropython setup.

**Reply by peterhinch** [reply]
That is a substantial library written in C. From a quick scan of the docs it doesn't seem to be supported by CircuitPython.

I can imagine that porting it to MP would be a big job. I can't imagine something that big and specialised being included in standard builds. Consequently to use it people would need the tools and skills to build firmware.

In use, to achieve decent quality, you'd need I2S output and an external I2S DAC. Given that you're stuck with external audio hardware, there seems little benefit over using a VS1053b, which also supports the CD quality FLAC format and works with a pure Python driver.

**Reply by Guustaaf** [reply]
Pretty good points. I ordered some different VS1053b boards. I will report on how it goes.

**Reply by rkompass** [reply]
Thanks for the clarifications. Given the VS1053b also has the amps for earphones it should be a good choice.

**Reply by Guustaaf** [reply]
I hooked up a VS1053 Arduino shield to a Pico and it works like a charm. Even though it is made for Arduino, the chip is not getting any more than 3.3v, so no level conversion required. The markings refer to Peter's IO nomenclature.
![wiring](https://user-images.githubusercontent.com/17580048/236656138-c0560cb6-e4b3-4a89-b1c3-708a07f208d0.jpg)

**Comment by Guustaaf** [comment]
Yes, that would really be great if it were also available for Micropython!

**Comment by robert-hh** [comment]
> This function is called from an interrupt, so it should be short and sweet. It's getting called 44,100 to 48,000 times per second for most MP3 files.

That's a serious limit for most boards supporting MicroPython. It would work with some MIMXRT boards like Teensy 4.x and some STM32x7 boards like the PYBD_SF6. But hardly with any ESP32 or RP2 device. And it requires a SD card as data source. So better use an external MP3 decoding chip.

**Reply by NewWheelTech** [reply]
They don't see to have any issues with using it with the RP2040:
https://learn.adafruit.com/mp3-playback-rp2040/pico-mp3

**Reply by mattytrentini** [reply]
Note the limitation in that learn guide: 
> Mono and stereo files less than 64kbit/s work, with sample rates from 8kHz to 24kHz. The RP2040 has a PWM output with 10 bits, so there's not much point in using high bit rates.

**Reply by rkompass** [reply]
Yes, accepted. But in a case where you have a 4-cm-diameter speaker without a case and just want to play some sounds this may be an attractive-by-being-simple solution. Given you had such a mp3 module.

**Reply by mattytrentini** [reply]
Oh, definitely, it's a really neat solution.

**Comment by scruss** [comment]
> CircuitPython has this great looking MP3 module.

Note that the CircuitPython player has some serious [limitations](https://learn.adafruit.com/mp3-playback-rp2040/pico-mp3#circuitpython-compatible-mp3-files-3101249) on the files it can play back:

>> Mono and stereo files less than 64kbit/s work, with sample rates from 8kHz to 24kHz

These aren't going to sound great, especially not through a 1-bit PWM output.

**Reply by peterhinch** [reply]
The distinction here is between making sounds and playing music. The VS1053b can play CD quality FLAC files. Polyphonic audio played via 1-bit PWM will produce an assortment of spurious intermodulation frequencies. Fine for model train whistles but for a cello concerto - not so good.

**Reply by scruss** [reply]
> Polyphonic audio played via 1-bit PWM will produce an assortment of spurious intermodulation frequencies. 

(Getting wildly off-topic here, and we've nothing like it for MicroPython, but: you can do wonderful things with that intermod. [4000AD | PROTODOME](https://protodome.bandcamp.com/album/4000AD) is an album created entirely using 1-bit audio on an AVR microcontroller, and formed part of the author's PhD in music composition.)

**Reply by peterhinch** [reply]
Extraordinary :)

**Comment by Jeff-Squared** [comment]
@Guustaaf @peterhinch hi guys, I am trying to complete an internet radio project of which I can already connect and download mp3 files to an SD card. Would I be correct in assuming that if I had a VS1053 MP3 shield like the one pictured by Guustaaf in this thread that I could just play the mp3 files directly to the output using SPI.

Stumbling upon this thread I was also interested to find the link to peterhinch's VS1053 device driver module and I have downloaded the zip file from git hub, I am sure I can make use of this driver but before I get too deep into that I am hoping someone can tell me that all I need is SPI.

Regarding boards I see Adafruit have a breakout for ~$25 Ali Express for as low as ~$3 and many more if you Google, but to split the difference I will order a shield at ~$15 that looks just like the one pictured here, mainly because I can get it within 5 days with free shipping.

Whatever happens I will post back here with feedback. BTW my project is based around the Arduino Nano ESP32 with Micropython.

**Comment by peterhinch** [comment]
[The docs](https://github.com/peterhinch/micropython-vs1053/tree/master) tell you all you want to know. Follow the link to the driver you want to use for wiring details.

**Reply by Jeff-Squared** [reply]
@peterhinch thank you, I did download a zip of the code yesterday and today I see there is even more information on your git hub page which is going to be helpful. Some of what I see is a little over my head but I think I can figure it out with perhaps a few prompts down the road.

My initial target is to run the synchronous  version of the driver with your provided example that uses the SD card and then look into replacing the SD card with the radio stream.

Initially I thought just give it an SPI signal but it seems like there are also some signals that go module ---> mcu (dreq for example) which could be 5v so that is something else I need to be aware of (ESP32 is a 3.3v mcu).

If we can put a radio stream through one of these modules using your driver that might create envy, has anyone seen an example of this?

thanks again.

**Reply by Guustaaf** [reply]
The VS1053 MP3 shield takes 5v power, but the chip itself runs on 3.3v so you should have no problems interfacing it directly with a 3.3v mcu. Please verify before hooking up.
Please explain how the concept of "envy" relates to this.

**Reply by Jeff-Squared** [reply]
@Guustaaf I was hoping that would be the case with the voltage levels and I will certainly verify as I go.

Perhaps my last sentence was not written well, it was just a few random thoughts I had and there was no intended emphasis on the word envy. 

regards and thank you

**Comment by Jeff-Squared** [comment]
@peterhinch @Guustaaf I put this project to one side for 3 months, for various reasons, but revisited it a couple of weeks ago. I have to thank you guys the results are outstanding and the audio quality is really good.

I created one project as an internet radio and a second as a MP3 music player. Both projects were built with a nano ESP32, an Arduino MP3 VS 1053 shield (pictured above) and Peters drivers vs1053_syn.py and sdcard.py. I plugged a $10 BT transmitter into the SPK jack of the shield which gives me the option to listen with any bluetooth device I have.

I still have to mount the electronics into project boxes and have yet to include a pot for volume. Lastly I think I would have been better off with Peters vs1053_asyn.py which I may look into further at some point.  

Thanks again I can't believe how good a micropython project this turned out to be.

**Comment by yingshaoxo** [comment]
Can anyone understand this mp3 decoder in c? It is written for micropython:  https://github.com/labplus-cn/mpython/blob/master/port/drivers/codec/audio_player.c#L157

According to my observation, it is using a mp3 decoder called 'helix': https://github.com/labplus-cn/mpython/tree/master/port/drivers/codec/codec_lib/helix/src

Hope it solves problems.

But don't put it into core micropython library, it is kind of big. A wav decoder is enough just like cpython did.
